### PR TITLE
WIP: Data grid render tile

### PR DESCRIPTION
**Pull request by attitude**
Adds ability to render tiles in the data grid

- [x] Grid UI component
- [x] Card UI component 
- [x] Update static renderer
- [x] Sandbox example

Data grid tile preview

---
This change is [Reviewable](https://reviewable.io/reviews/contember/admin/315)

### CODE REVIEW

**Review by matej21 — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 6 of 6 files at r5, 2 of 4 files at r6, 5 of 24 files at r10, 23 of 23 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/contember/admin/315)* status: all files reviewed, 17 unresolved discussions (waiting on @attitude)

___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/base/DataGridContainer.tsx` line 135 at r11](https://reviewable.io/reviews/contember/admin/315#-N1ngZPQ3vq3ATJ1bayI:-N1ngZPQ3vq3ATJ1bayJ:b-oibmwm) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/base/DataGridContainer.tsx#L135)):*
> ```XML
> 					</Stack>
> 				</LayoutPageStickyContainer>
> 				{tile && layout === 'tiles'
> ```

this component is getting too big, could you please split it into multiple components?
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 1 at r11](https://reviewable.io/reviews/contember/admin/315#-N1nh7du8dVa51rlWA3Y:-N1nh7du8dVa51rlWA3Z:b-mg8650) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L1)):*
> ```XML
> import { Component, Field, useEntity } from '@contember/binding'
> ```

bad indent in whole file
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 3 at r11](https://reviewable.io/reviews/contember/admin/315#-N1nh9XTEgjN3mWJ7dSW:-N1nh9XTEgjN3mWJ7dSX:bjs4mhx) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L3)):*
> ```XML
> import { Component, Field, useEntity } from '@contember/binding'
> import { Card, CardProps } from '@contember/ui'
> import { CheckboxButtonProps } from '@contember/ui/src/components/Forms/Checkbox/CheckboxButton'
> ```

bad import
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 7 at r11](https://reviewable.io/reviews/contember/admin/315#-N1ngsw1B6bXxcAdDDqA:-N1ngsw1B6bXxcAdDDqB:b-lt8ukv) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L7)):*
> ```XML
> import { Link, LinkProps } from '../../../../../routing'
> 
> export interface DataGridTileProps extends
> ```

use type instad of interface for complex types
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 32 at r11](https://reviewable.io/reviews/contember/admin/315#-N1nhZiS7LQqfnIq-DiQ:-N1nhZiTB-H1hT2nIbie:b98py7f) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L32)):*
> ```XML
>   const cardProps = useMemo(() => ({
>     ...props,
>     src: src as string,
> ```

use `getField<string>` instead
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 51 at r11](https://reviewable.io/reviews/contember/admin/315#-N1nhfqHDc8AkBkhMF55:-N1nhfqHDc8AkBkhMF56:b-q0qgt7) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L51)):*
> ```XML
>   ...props
> }) => {
>   console.log({ thumbnailField })
> ```

remove debug output
___
*[`packages/ui/src/components/Card/Card.tsx` line 10 at r5](https://reviewable.io/reviews/contember/admin/315#-N1ndim1AWp4LVUbGEk9:-N1ndim1AWp4LVUbGEkA:b-6scnuo) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L10)):*
> ```XML
> 
> type LinkCompatibleProps<E extends HTMLDivElement | HTMLAnchorElement> = E extends HTMLDivElement
>   ? {
> ```

you accept props from HTMLAnchorElement but not from div?
___
*[`packages/ui/src/components/Card/Card.tsx` line 11 at r5](https://reviewable.io/reviews/contember/admin/315#-N1ndSYBCnvfH1PnxmQJ:-N1ndSYBCnvfH1PnxmQK:b-se0he7) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L11)):*
> ```XML
> type LinkCompatibleProps<E extends HTMLDivElement | HTMLAnchorElement> = E extends HTMLDivElement
>   ? {
>     href?: string
> ```

I think, you've wanted to use `href?: never`, right?
___
*[`packages/ui/src/components/Card/Card.tsx` line 18 at r5](https://reviewable.io/reviews/contember/admin/315#-N1nd3bxDM4UpKCgLeYt:-N1nd3bxDM4UpKCgLeYu:btnw5hv) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L18)):*
> ```XML
>     href: string
>     active: boolean
>     onClick: (e?: MouseEvent<HTMLAnchorElement>) => void
> ```

in which case is `e` undefined?
___
*[`packages/ui/src/components/Card/Card.tsx` line 22 at r5](https://reviewable.io/reviews/contember/admin/315#-N1neT2WDNu6r4w_RwQB:-N1neT2WDNu6r4w_RwQC:b-c9xxu2) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L22)):*
> ```XML
> 
> export type CardProps<E extends HTMLDivElement | HTMLAnchorElement> =
>   Omit<VisuallyDependentControlProps, 'active' | 'type'> &
> ```

preferred formatting of complex types is:
```
export type Foo = 
  & Bar
  & Lorem
  & {
    inpum: string
  }
```
___
*[`packages/ui/src/components/Card/Card.tsx` line 43 at r5](https://reviewable.io/reviews/contember/admin/315#-N1nd9l27VLYp5OfN-uQ:-N1nd9l27VLYp5OfN-uR:b-w6dv30) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L43)):*
> ```XML
>   const Container = isLink ? 'a' : 'div'
> 
>   return <Container
> ```

cs: wrap into `()`
___
*[`packages/ui/src/components/Card/Card.tsx` line 52 at r5](https://reviewable.io/reviews/contember/admin/315#-N1ne4GkBbXXb-x5dZL7:-N1ne4GkBbXXb-x5dZL8:b42tv1h) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L52)):*
> ```XML
>         ? ['hover', 'press', 'focus']
>         : undefined
>       )?.map(feature => toFeatureClass(feature, true)),
> ```

it is hard to read, I would prefer
```
toFeatureClass('hover', isLink),
toFeatureClass('press', isLink),
toFeatureClass('focus', isLink),
```
___
*[`packages/ui/src/components/Card/Card.tsx` line 55 at r5](https://reviewable.io/reviews/contember/admin/315#-N1ndcWlB2sQWN4BabMM:-N1ndcWlB2sQWN4BabMN:b-nczfxp) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L55)):*
> ```XML
>       useInputClassName(props as VisuallyDependentControlProps),
>     )}
>   >
> ```

in props you accept `Omit<AnchorHTMLAttributes<HTMLAnchorElement>, 'href'>`, so I would expect `{..props}` here.
___
*[`packages/ui/src/components/Card/Card.tsx` line 59 at r5](https://reviewable.io/reviews/contember/admin/315#-N1neJp49-tWOzLpxH5c:-N1neJp49-tWOzLpxH5d:b-fkne9h) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L59)):*
> ```XML
>       <div
>         className={`${componentClassName}-thumbnail`}
>         style={{ backgroundImage: src ? `url(${src})` : undefined }}
> ```

I would consider escaping the `src`
___
*[`packages/ui/src/components/Grid/Grid.tsx` line 22 at r10](https://reviewable.io/reviews/contember/admin/315#-N1nf9cB1_-4pKYA9iuM:-N1nf9cB1_-4pKYA9iuN:b-a9nu0f) ([raw file](https://github.com/contember/admin/blob/9d7cd485a51b2394c05aa3ff9f2124a48cae85c2/packages/ui/src/components/Grid/Grid.tsx#L22)):*
> ```XML
>     }: GridProps, forwardedRef) => {
>       if (columnWidth < 0) {
>         throw new Error('Column width must be a non-negative number larger than 0')
> ```

error message contains pleonasm :)
___
*[`packages/ui/src/components/Grid/Grid.tsx` line 31 at r10](https://reviewable.io/reviews/contember/admin/315#-N1nfC_45KG3SlPC4sXK:-N1nfC_45KG3SlPC4sXL:b-k4b22a) ([raw file](https://github.com/contember/admin/blob/9d7cd485a51b2394c05aa3ff9f2124a48cae85c2/packages/ui/src/components/Grid/Grid.tsx#L31)):*
> ```XML
>       }) : undefined, [columnWidth])
> 
>       return <div
> ```

wrap into `()`
___
*[`packages/ui/src/utils/useFallbackRef.ts` line 6 at r11](https://reviewable.io/reviews/contember/admin/315#-N1ngJiHDdxD_Jc9m9kg:-N1ngJiI2ppaCELZ7a2b:bjsgjqk) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/ui/src/utils/useFallbackRef.ts#L6)):*
> ```XML
> 	const innerRef = useRef<E>(null)
> 
>   return forwardedRef ?? innerRef
> ```

bad indent

**Review by attitude — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/contember/admin/315)* status: 32 of 45 files reviewed, 17 unresolved discussions (waiting on @attitude and @matej21)

___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/base/DataGridContainer.tsx` line 135 at r11](https://reviewable.io/reviews/contember/admin/315#-N1ngZPQ3vq3ATJ1bayI:-N1ryJFo5vLOHQuKsBfJ:b-896fix) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/base/DataGridContainer.tsx#L135)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

this component is getting too big, could you please split it into multiple components?
</blockquote></details>

Done.
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 1 at r11](https://reviewable.io/reviews/contember/admin/315#-N1nh7du8dVa51rlWA3Y:-N1qunBu9emgEViOBJPp:b-896fix) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L1)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

bad indent in whole file
</blockquote></details>

Done.
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 3 at r11](https://reviewable.io/reviews/contember/admin/315#-N1nh9XTEgjN3mWJ7dSW:-N1qvDoi8jbXSFoemRv1:b-896fix) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L3)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

bad import
</blockquote></details>

Done.
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 7 at r11](https://reviewable.io/reviews/contember/admin/315#-N1ngsw1B6bXxcAdDDqA:-N1qvQZp8alA2Arf_MpC:b-896fix) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L7)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

use type instad of interface for complex types
</blockquote></details>

Done.
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 32 at r11](https://reviewable.io/reviews/contember/admin/315#-N1nhZiS7LQqfnIq-DiQ:-N1qvZUk-Ds5J7ULUP8N:b-896fix) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L32)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

use `getField<string>` instead
</blockquote></details>

Done.
___
*[`packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx` line 51 at r11](https://reviewable.io/reviews/contember/admin/315#-N1nhfqHDc8AkBkhMF55:-N1qvfgXAO2ISYdPOOH9:b-896fix) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/admin/src/components/bindingFacade/collections/DataGrid/grid/DataGridTile.tsx#L51)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

remove debug output
</blockquote></details>

Done.
___
*[`packages/ui/src/components/Card/Card.tsx` line 10 at r5](https://reviewable.io/reviews/contember/admin/315#-N1ndim1AWp4LVUbGEk9:-N1rX9oM7G9Lc_WYX2ck:b-896fix) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L10)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

you accept props from HTMLAnchorElement but not from div?
</blockquote></details>

Done.
___
*[`packages/ui/src/components/Card/Card.tsx` line 11 at r5](https://reviewable.io/reviews/contember/admin/315#-N1ndSYBCnvfH1PnxmQJ:-N1qw1K483IN2j6anRYh:bb77ufg) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L11)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

I think, you've wanted to use `href?: never`, right?
</blockquote></details>

Yes.
___
*[`packages/ui/src/components/Card/Card.tsx` line 18 at r5](https://reviewable.io/reviews/contember/admin/315#-N1nd3bxDM4UpKCgLeYt:-N1qxLUS6TiKiwyXMkbh:b-yt4xc9) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L18)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

in which case is `e` undefined?
</blockquote></details>

Didn't explore. Exists to comply with interface of `InnerRoutingLinkProps` (packages/admin/src/routing/RoutingLink.tsx).
___
*[`packages/ui/src/components/Card/Card.tsx` line 22 at r5](https://reviewable.io/reviews/contember/admin/315#-N1neT2WDNu6r4w_RwQB:-N1qyE74629CZ_6Wo7jB:b-896fix) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L22)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

preferred formatting of complex types is:
```
export type Foo = 
  & Bar
  & Lorem
  & {
    inpum: string
  }
```
</blockquote></details>

Done.
___
*[`packages/ui/src/components/Card/Card.tsx` line 43 at r5](https://reviewable.io/reviews/contember/admin/315#-N1nd9l27VLYp5OfN-uQ:-N1rXRTuDbitN_cHC7Tp:b-896fix) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L43)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

cs: wrap into `()`
</blockquote></details>

Done.
___
*[`packages/ui/src/components/Card/Card.tsx` line 52 at r5](https://reviewable.io/reviews/contember/admin/315#-N1ne4GkBbXXb-x5dZL7:-N1qyxDzC0nWVIAdPifb:b-896fix) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L52)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

it is hard to read, I would prefer
```
toFeatureClass('hover', isLink),
toFeatureClass('press', isLink),
toFeatureClass('focus', isLink),
```
</blockquote></details>

Done.
___
*[`packages/ui/src/components/Card/Card.tsx` line 55 at r5](https://reviewable.io/reviews/contember/admin/315#-N1ndcWlB2sQWN4BabMM:-N1qzOm-0LHZeYOEku_-:b-896fix) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L55)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

in props you accept `Omit<AnchorHTMLAttributes<HTMLAnchorElement>, 'href'>`, so I would expect `{..props}` here.
</blockquote></details>

Done.
___
*[`packages/ui/src/components/Card/Card.tsx` line 59 at r5](https://reviewable.io/reviews/contember/admin/315#-N1neJp49-tWOzLpxH5c:-N1qzZbb3dUHFyEfQnzf:b-896fix) ([raw file](https://github.com/contember/admin/blob/94c29849155d76a15bfd42c4f43390c20ec40364/packages/ui/src/components/Card/Card.tsx#L59)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

I would consider escaping the `src`
</blockquote></details>

Done.
___
*[`packages/ui/src/components/Grid/Grid.tsx` line 22 at r10](https://reviewable.io/reviews/contember/admin/315#-N1nf9cB1_-4pKYA9iuM:-N1rWnBeEUgncQ02JPP9:b-896fix) ([raw file](https://github.com/contember/admin/blob/9d7cd485a51b2394c05aa3ff9f2124a48cae85c2/packages/ui/src/components/Grid/Grid.tsx#L22)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

error message contains pleonasm :)
</blockquote></details>

Done.
___
*[`packages/ui/src/components/Grid/Grid.tsx` line 31 at r10](https://reviewable.io/reviews/contember/admin/315#-N1nfC_45KG3SlPC4sXK:-N1rX-bB1NvMGtCQvcPp:b-896fix) ([raw file](https://github.com/contember/admin/blob/9d7cd485a51b2394c05aa3ff9f2124a48cae85c2/packages/ui/src/components/Grid/Grid.tsx#L31)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

wrap into `()`
</blockquote></details>

Done.
___
*[`packages/ui/src/utils/useFallbackRef.ts` line 6 at r11](https://reviewable.io/reviews/contember/admin/315#-N1ngJiHDdxD_Jc9m9kg:-N1rX7E8DpIpCpwpgmKg:b-896fix) ([raw file](https://github.com/contember/admin/blob/b3ad32e20847ca1e784a292abd2fcea029248b07/packages/ui/src/utils/useFallbackRef.ts#L6)):*
<details><summary><i>Previously, matej21 (David Matějka) wrote…</i></summary><blockquote>

bad indent
</blockquote></details>

Done.

**Review by matej21 — APPROVED** [review]
> Reviewed 13 of 13 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/contember/admin/315)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @attitude)

**Review by matej21 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/contember/admin/315)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @attitude)

File: packages/admin-sandbox/admin/pages/article.tsx [hunk 1]
```diff
@@ -30,36 +27,45 @@ const stateOptions = {
...
-			</DataBindingProvider>
-		</GenericPage>
-	)
-}
+export const List = () => <DataGridPage
+	entities="Article"
+	itemsPerPage={20}
+	tile={<DataGridTile
+		to={`article/edit(id: $entity.id)`}
+		thumbnailField="image.url"
```
Comment (VojtaStanek) [inline packages/admin-sandbox/admin/pages/article.tsx:35]:
```suggestion
		thumbnail={<Image field="image.url />}
```

**Comment by VojtaStanek** [comment]
How is this compatible with custom image components?

**Pull request merged**